allennlp/data/iterators/data_iterator.py

```diff
@@ -35,13 +38,20 @@ def __call__(self,
         shuffle : ``bool``, optional (default=``True``)
             If ``True``, we will shuffle the instances in ``dataset`` before constructing batches
             and iterating over the data.
+        cuda_device : ``int``
```

Comment:
I don't _think_ we should do device stuff here. If we do, I think it will be harder to switch to using multiple GPUs.

The problem is the following: to do multi-gpu stuff in pytorch, your batch is split along the batch dimension and pytorch handles sending it to the GPUs. This definitely needs to happen in trainer, so it makes doing any device manipulation here not that feasible. The annoying thing is that this basically means we need to have a recursive `tensor_to_gpu` function ....

Reply:
What if we had a `MultiGpuDataIterator` that rotated the devices used for each batch, and the `Trainer` picked out as many batches as it needed?  Would that solve the problem?  That would be a pretty minimal change from this API.

Either way, I think that's an issue to solve when we implement multi-gpu stuff.  I don't think there's anything I can change in this PR to make that easier later - we need to know what the multi-gpu solution is actually going to look like before we can make the right change to how device info is passed around.